packages/block-library/src/editor.scss

```diff
@@ -53,7 +53,7 @@
 
 // Provide every block with a default base margin. This margin provides a consistent spacing
 // between blocks in the editor.
-.editor-styles-wrapper [data-block] {
+.editor-styles-wrapper .block-editor-block-list__block.has-spacing > .block-editor-block-list__block-edit > [data-block] {
```

Comment:
As a quick preface, many of these review items happened on a separate PR where I did not have the full context of what you're trying to do here. Now that I do have that context, I applaud you for taking on the task. It's difficult, and many review points need to be thought about in a different way, but I absolutely see the purpose and want to make myself available to help if I can. Notably around spacing I have a great deal of institutional knowledge that might help there. 

So we do need to avoid specificity whenever possible. I'd like to try and zoom out a bit here. 

1. We do want to provide a baseline margin to most blocks.
2. But not all, blocks should be able to opt out.

If that is the goal, can we do something like `[data-has-spacing]`? For #17267 I was thinking whether it might make sense to have a `[data-block-container]` as well, to help ambiguate.

The benefit of using these data attributes is that they are entirely optional, and they are added to the innermost div they can be, which means we can avoid the `>` selector which is very hard to override in CSS.